### ISSUE TITLE
Confusing behaviour with `.over`

### ISSUE THREAD

**Issue by guigzzz** [pola-rs/polars#9749]
### Polars version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pypi.org/project/polars/) of Polars.

### Issue description

I don't actually know if this is a bug, but I would expect `v1` and `v2`  to be equal.
Should I expect the `.over` to apply to the `.mean()` in the ternary expression ?
If I sort the dataframe first and then apply the ternary expression it works, so my gut feeling is this a bug.

### Reproducible example

```python
i = pl.col("Int")
df = pl.DataFrame([("A", 0), ("B", 2), ("A", 1), ("B", 3)], schema=["Letter", "Int"])
df.select(
    "Letter",
    "Int",
    pl.when(i > 0).then(i.mean()).otherwise(i).over("Letter").alias("v1"),
    pl.when(i > 0).then(i.mean().over("Letter")).otherwise(i).alias("v2"),
).sort("Letter").with_columns(
    pl.when(i > 0).then(i.mean()).otherwise(i).over("Letter").alias("v3")
)
```

### Expected behavior

`v1`, `v2`, and `v3` should be exactly equal.

### Installed versions

```
--------Version info---------
Polars:              0.18.6
Index type:          UInt32
Platform:            Linux-5.10.16.3-microsoft-standard-WSL2-x86_64-with-glibc2.29
Python:              3.8.10 (default, Mar 13 2023, 10:26:41) 
[GCC 9.4.0]

----Optional dependencies----
adbc_driver_sqlite:  <not installed>
connectorx:          <not installed>
deltalake:           <not installed>
fsspec:              <not installed>
matplotlib:          <not installed>
numpy:               1.24.2
pandas:              1.5.3
pyarrow:             11.0.0
pydantic:            <not installed>
sqlalchemy:          <not installed>
xlsx2csv:            <not installed>
xlsxwriter:          <not installed>
```

**Comment by avimallu** [comment]
There's an error message you get (twice) when you attempt this transformation (not sure if you did):

`The predicate '[(col("Int")) > (0)]' in 'when->then->otherwise' is not a valid aggregation and might produce a different number of rows than the groupby operation would. This behavior is experimental and may be subject to change`

The key point here is that an `over` statement is just a `groupby` on the operation **with an `explode`** to put the **aggregated** row values back in their normal place. Let's see what happens in a `groupby`:

```py
(
    df
    .groupby("Letter")
    .agg(
        pl.when(i > 0)
        .then(i.mean())
        .otherwise(i))
)
```

gives:

```py
shape: (2, 2)
┌────────┬────────────┐
│ Letter ┆ Int        │
│ ---    ┆ ---        │
│ str    ┆ list[f64]  │
╞════════╪════════════╡
│ A      ┆ [0.0, 0.5] │
│ B      ┆ [2.5, 2.5] │
└────────┴────────────┘
```

Now because the aggregation condition for `when` is not proper, that is, it returns two values per row rather than one, Polars doesn't know what to do with the 4 values it gets after the explode. When you sort the dataframe later (output `v3`), you _don't_ actually get the correct output (as _you_ perceive it), but the same output exploded back to a DataFrame that has the `Letter` column in the same order as `groupby` would have provided.

How is `v2` different though? In `v2`, you _don't_ have an aggregation inside the `then`. You have a `mean` inside an `over` that _already_ produces a **single** output per group (i.e. per `Letter`), and that Polars is able to correctly `explode` to accurate row positions.

Hope that makes sense. Take this an a response from a Polars user, not a developer.
TL;DR: not a bug since you explicitly get a warning about "improper" aggregation, followed by an "experimental" label :)

**Comment by cmdlineluser** [comment]
I think `v1` and `v3` are going to raise in the future:

> Maybe we should raise an exception when .over is used without a window?

https://github.com/pola-rs/polars/issues/8692
